### PR TITLE
Translate Quint `foldr`

### PR DESCRIPTION
**Pull request by thpani**
Towards #2437, #2479

As discussed synchronously, add an in-code translation for Quint `foldr`.

I will rebase onto `main` once #2483 is reviewed & merged.

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] ~Documentation added for any new functionality~
- [ ] ~[Entries added to `./unreleased/`][changelog format] for any new functionality~

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

**Review by konnov — APPROVED** [review]
Looks good to me! I hope that the preprocessor would not choke on all these lambdas. But there is only one way to learn :)

File: tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala [hunk 1]
```diff
@@ -339,7 +339,43 @@ class Quint(moduleData: QuintOutput) {
                 })(quintArgs)
         // TODO: list operations requiring rewiring  https://github.com/informalsystems/apalache/issues/2437
         case "range" => null
-        case "foldr" => null
+        case "foldr" =>
+          // TODO: unique names for LET-INs!
```
[the remark this answers is not in the document]
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala]: Done in [6664ae6](https://github.com/informalsystems/apalache/pull/2488/commits/6664ae65eca3257bb6684c549bd70cd8836f9914)

**Review by shonfeder — COMMENTED** [review]
Generally looks great!

**Review by shonfeder — APPROVED** [review]

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala [hunk 2]
```diff
@@ -396,6 +396,21 @@ class TestQuintEx extends AnyFunSuite {
...
+        "LET __QUINT_LAMBDA2(__quint_var1) ≜ __quint_var0[(__QUINT_LAMBDA1() - __quint_var1) + 1] IN __QUINT_LAMBDA2"
+      val reverseDecl =
+        s"__QUINT_LAMBDA3(__quint_var0) ≜ LET ${slenDecl} IN Sequences!SubSeq(Apalache!MkSeq(ApalacheInternal!__ApalacheSeqCapacity(__quint_var0), ${get_ith}), 1, __QUINT_LAMBDA1())"
+      val map =
+        "LET __QUINT_LAMBDA4(__quint_var3, __quint_var2) ≜ LET __QUINT_LAMBDA0(acc, n) ≜ isum(n, acc) IN __QUINT_LAMBDA0(__quint_var2, __quint_var3) IN __QUINT_LAMBDA4"
+      val reversedSeq = "__QUINT_LAMBDA3(<<1, 2, 3>>)"
+      s"LET ${reverseDecl} IN Apalache!ApaFoldSeqLeft(${map}, 0, ${reversedSeq})"
+    }
+    assert(convert(Q.app("foldr", Q.intList, Q._0, Q.accumulatingOpp)) == expected)
+  }
```
[the remark this answers is not in the document]
Comment (thpani) [reply tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:412]: Hmmm, I've had it in a single line, but then it becomes (even more) unreadable.

WDYT?

File: tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala [hunk 3]
```diff
@@ -343,7 +343,42 @@ class Quint(moduleData: QuintOutput) {
...
-        case "foldr" => null
+        case "foldr" =>
+          quintArgs =>
+            ternaryApp(opName,
+                (seq, init, op) => {
+                  val seqType = Quint.typeToTlaType(types(quintArgs(0).id).typ)
+                  val elemType = seqType match {
+                    case SeqT1(elem) => elem
+                    case invalidType => throw new QuintIRParseError(s"sequence ${seq} has invalid type ${invalidType}")
+                  }
```
[the remark this answers is not in the document]
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala:354]: Totally. I wanted to put in most of the translation first, to see how we can best refactor!

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:412]: yeah, makes sense. I guess readability is more important here.  happy to have it stay as is.

**Pull request merged**